sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuAvroScan.scala

```diff
@@ -151,92 +151,276 @@ case class GpuAvroPartitionReaderFactory(
+    partitionSchema: StructType,
+    filters: Array[Filter],
+    @transient rapidsConf: RapidsConf,
+    options: AvroOptions,
+    metrics: Map[String, GpuMetric],
+    queryUsesInputFile: Boolean)
+  extends MultiFilePartitionReaderFactoryBase(sqlConf, broadcastedConf, rapidsConf) {
 
-  def filterBlocks(partFile: PartitionedFile): AvroBlockMeta = {
+  private val debugDumpPrefix = rapidsConf.avroDebugDumpPrefix
+  private val ignoreMissingFiles = sqlConf.ignoreMissingFiles
+  private val ignoreCorruptFiles = sqlConf.ignoreCorruptFiles
+
+  private val numThreads = rapidsConf.avroMultiThreadReadNumThreads
+  private val maxNumFileProcessed = rapidsConf.maxNumAvroFilesParallel
+
+  // Disable coalescing reading until it is supported.
+  override val canUseCoalesceFilesReader: Boolean = false
+
+  override val canUseMultiThreadReader: Boolean = rapidsConf.isAvroMultiThreadReadEnabled
+
+  /**
+   * File format short name used for logging and other things to uniquely identity
+   * which file format is being used.
+   */
+  override final def getFileFormatShortName: String = "AVRO"
+
+  /**
+   * Build the PartitionReader for cloud reading
+   */
+  override def buildBaseColumnarReaderForCloud(
+      files: Array[PartitionedFile],
+      conf: Configuration): PartitionReader[ColumnarBatch] = {
+    val filterHandler = AvroFileFilterHandler(conf, options)
+    new GpuMultiFileCloudAvroPartitionReader(conf, files, readDataSchema, partitionSchema,
+      maxReadBatchSizeRows, maxReadBatchSizeBytes, numThreads, maxNumFileProcessed,
+      debugDumpPrefix, filters, filterHandler, metrics, ignoreMissingFiles, ignoreCorruptFiles)
+  }
+
+  /**
+   * Build the PartitionReader for coalescing reading
+   *
+   * @param files files to be read
+   * @param conf  the configuration
+   * @return a PartitionReader of coalescing reading
+   */
+  override def buildBaseColumnarReaderForCoalescing(
+      files: Array[PartitionedFile],
+      conf: Configuration): PartitionReader[ColumnarBatch] = {
+    throw new UnsupportedOperationException
+  }
+
+}
+
+/** A trait collecting common methods across the 3 kinds of avro readers */
+trait GpuAvroReaderBase extends Arm with Logging { self: FilePartitionReaderBase =>
 
-    def passSync(blockStart: Long, position: Long): Boolean = {
-      blockStart >= position + SYNC_SIZE
+  def debugDumpPrefix: String
+
+  def readDataSchema: StructType
+
+  /**
+   * Send a host buffer to GPU for decoding.
+   * The input hostBuf will be closed after returning, please do not use it anymore.
+   * 'splits' is used only for debugging.
+   */
+  protected def sendToGpu(
+      hostBuf: HostMemoryBuffer,
+      bufSize: Long,
+      splits: Array[PartitionedFile]): Option[ColumnarBatch] = {
+    withResource(hostBuf) { _ =>
+      if (bufSize == 0) {
+        None
+      } else {
+        // Dump buffer for debugging when required
+        dumpDataToFile(hostBuf, bufSize, splits, Option(debugDumpPrefix), Some("avro"))
```

Comment:
Nit: Rather than create a new Option object every time this is called, why not create it once when the RapidsConf value is read and pass that along to be reused?  Similarly we're creating an "avro" Option object here that is almost never used, would be good to cache this and reuse to avoid unnecessary garbage creation.

Reply:
Done